lib/tracing-limit/src/lib.rs

```diff
@@ -1,210 +1,297 @@
-                            }
-                        }
-                    }
-
-                    return false;
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        let metadata = event.metadata();
+
+        // if the event is not rate limited, just pass through
+        if !is_limited(metadata) {
+            return self.inner.on_event(event, ctx);
+        }
+
+        let mut limit_visitor = LimitVisitor::default();
+        event.record(&mut limit_visitor);
+
+        let component_name = {
+            let mut scope = ctx
+                .lookup_current()
+                .into_iter()
+                .flat_map(|span| span.from_root().chain(std::iter::once(span)));
+
+            scope.find_map(|span| {
+                let extensions = span.extensions();
+                extensions
+                    .get::<RateLimitedSpanKeys>()
+                    .and_then(|fields| fields.component_name.clone())
+            })
+        };
+
+        let id = RateKeyIdentifier {
+            callsite: metadata.callsite(),
+            component_name,
+        };
+
+        let mut state = self.events.entry(id).or_insert(State {
+            start: Instant::now(),
+            count: 0,
+            // if this is None, then a non-integer was passed as the rate limit
+            limit: limit_visitor
+                .limit
+                .expect("unreachable; if you see this, there is a bug"),
+            message: limit_visitor
+                .message
+                .unwrap_or_else(|| event.metadata().name().into()),
+        });
+
+        let prev = state.count;
+        state.count += 1;
+
+        //check if we are still rate limiting
+        if state.start.elapsed().as_secs() < state.limit.get() {
+            // check and increment the current count
+            // if 0: this is the first message, just pass it through
+            // if 1: this is the first rate limited message
+            // otherwise supress it until the rate limit expires
+            match prev {
+                0 => self.inner.on_event(event, ctx),
+                1 => {
+                    // output first rate limited log
+                    let message =
+                        format!("Internal log [{}] is being rate limited.", state.message);
+                    self.create_event(&ctx, metadata, message, state.limit);
                 }
+                _ => {}
+            }
+        } else {
+            // done rate limiting
+
+            // output a message if any events were rate limited
+            if prev > 1 {
+                let message = format!(
+                    "Internal log [{}] has been rate limited {} times.",
+                    state.message,
+                    prev - 1
+                );
+
+                self.create_event(&ctx, metadata, message, state.limit);
             }
+            self.inner.on_event(event, ctx);
+            state.start = Instant::now();
+            state.count = 1;
```

Comment:
I think this should be set to 0? If we're resetting the interior state back to default then we need to be zero, unless I'm missing something. Else the next time we come back around here the count will be set to 2, on account of line 150?

Reply:
I also found this a bit confusing :)

We want to set it to 1 here because we just emitted the event above so the next one that comes in should be rate limited (thus we want it to be 2 the next time around).

Reply:
Aaaaah, yes. Makes perfect sense. I think if you found it confusing and I found it confusing I'd appreciate a comment explaining this.